x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ClusterComputeHandler.java

```
@@ -74,69 +76,107 @@ void startComputeOnRemoteCluster(
-            var resp = finalResponse.get();
-            return Objects.requireNonNullElseGet(resp, () -> new ComputeResponse(profiles));
-        }))) {
-            ExchangeService.openExchange(
-                transportService,
-                cluster.connection,
-                childSessionId,
-                queryPragmas.exchangeBufferSize(),
-                esqlExecutor,
-                EsqlCCSUtils.skipUnavailableListener(
-                    computeListener.acquireAvoid(),
-                    executionInfo,
-                    clusterAlias,
-                    EsqlExecutionInfo.Cluster.Status.SKIPPED
-                ).delegateFailureAndWrap((l, unused) -> {
-                    var listenerGroup = new RemoteListenerGroup(
-                        transportService,
-                        rootTask,
-                        computeListener,
-                        clusterAlias,
-                        executionInfo,
-                        l
-                    );
-
-                    var remoteSink = exchangeService.newRemoteSink(
-                        listenerGroup.getGroupTask(),
-                        childSessionId,
-                        transportService,
-                        cluster.connection
-                    );
+        final AtomicBoolean pagesFetched = new AtomicBoolean();
+        final AtomicReference<ComputeResponse> finalResponse = new AtomicReference<>();
+        listener = listener.delegateResponse((l, e) -> {
+            final boolean receivedResults = finalResponse.get() != null || pagesFetched.get();
+            if (receivedResults == false && EsqlCCSUtils.shouldIgnoreRuntimeError(executionInfo, clusterAlias, e)) {
+                EsqlCCSUtils.markClusterWithFinalStateAndNoShards(executionInfo, clusterAlias, EsqlExecutionInfo.Cluster.Status.SKIPPED, e);
+                l.onResponse(List.of());
+            } else if (configuration.allowPartialResults()) {
```

Here I'm not sure about the logic of it - is we have allow_partial_results, does it mean that any failure is ignored now if we got any of the results? I.e. if we got some results, but then the query was cancelled - does it now considered success? Same for e.g. exception in query processing? 

I don't have any objections in principle for this approach, just noting it will then be timing-dependent - i.e. if the query is slow and you cancel it, it's an error, but if it's a bit faster and managed to send 1 row of results, it's a success. Same for exceptions - if the query manages to produce an exception, it's an error, if no results were delivered, but if it generated some results beforehand, it's ok. So the same query may be successful or failing depending on data. 

Again, this may be a valid approach, but I feel we need to spell out our expectations here explicitly. Maybe at least in the code comment somewhere in this area, ideally also in some kind of design document? Because if I'm not sure what's supposed to happen here, we can reasonably assume a lot of other people also wouldn't know what the expected logic here.

---

A similar situation can occur with allow_partial_results=false. I don't think we should comment on this case, as it wouldn't add anything and requiring future readers to think about it seems unnecessary.

---

But aren't there some exceptions that don't apply to allow_partial_results=true. Such as TaskCancelledException - if that's the case, don't we want the error to be "fatal" and return a top level error?

---

There are no special exceptions for allow_partial_results. We ignore all exceptions and return as many results as possible to the user. I think the same applies to the _search API. We don't need to check if the query is being cancelled. If a user cancels a query, whether we return partial results or fail is no longer important. However, if the query is being cancelled, we should respond quickly by not continuing the query to avoid wasting resources.

---

with allow_partial_results=false we would just proceed to `onFailure` so I am not sure how is it similar? Without partial results, failure on remote leads to overall failure (unless skip_un above intercepts it)? But with allow_partial_results=true we don't check which exception caused the failure if `receivedResults` is true - that's the part I do not understand.

---

f we have received results (either pages or a response, such as profiles) from a remote cluster, the skip_unavailable flag no longer applies because the cluster was not unavailable, but rather the query failed. The allow_partial_results flag, on the other hand, always applies, whether we receive results or not.

---

> The allow_partial_results flag, on the other hand, always applies, whether we receive results or not.

OK so that's what I was trying to figure out. Looks like the default mode with this flag (which is also set by default as I understand) is to ignore any (runtime) errors from remotes. This is a big change from how it was before, that's why I want it to be documented explicitly. Also, does this apply to local data too? Wouldn't it lead to cancellations being ignored (i.e. leading to not failing queries) too? I mean I don't exactly against this, it's just a big change in how we handle failures, and not sure people would realize there's such a big difference between 9.0 and 9.1 in how failures are processed by default. This particular patch does not enable the default of true, so everything is still the same but once we do, it changes a lot of assumptions. E.g. by default cancelling a query no longer results in query failure.

---

It defaults to false now, and switching the default to true would be a breaking change. We've been discussing this internally, and we believe the right default is allow_partial_results, as it gives users the best experience with ES|QL. We agreed to enable this by default in 8.19 and 9.1. Of course, we will document this properly and provide users a way to fall back to the previous behavior if needed.